src/Voron/Impl/StorageEnvironmentState.cs

```diff
@@ -1,32 +1,47 @@
 ﻿using System;
+using System.Collections.Generic;
 using Voron.Data.BTrees;
+using Voron.Impl.Journal;
 
 namespace Voron.Impl
 {
     public class StorageEnvironmentState
     {
-        public TreeMutableState Root { get; set; }
-        public StorageEnvironmentOptions Options { get; set; }
+        public TreeMutableState Root;
+        public StorageEnvironmentOptions Options;
+        
+        public long TransactionCounter;
+        public List<JournalSnapshot> SnapshotCache;
+        public Dictionary<int, PagerState> PagerStatesAllScratchesCache;
+        public object ExternalState;
 
         public long NextPageNumber;
 
-        public StorageEnvironmentState() { }
+        private StorageEnvironmentState()
+        {
+        }
 
         public StorageEnvironmentState(Tree root, long nextPageNumber)
         {
             if (root != null)
                 Root = root.State;
             NextPageNumber = nextPageNumber;
+            SnapshotCache = new List<JournalSnapshot>();
+            PagerStatesAllScratchesCache = new Dictionary<int, PagerState>();
         }
 
         public StorageEnvironmentState Clone()
         {
             return new StorageEnvironmentState
-                {
-                    Root = Root?.Clone(),
-                    NextPageNumber = NextPageNumber,
-                    Options = Options
-                };
+            {
+                Root = Root?.Clone(),
+                NextPageNumber = NextPageNumber,
+                Options = Options,
+                SnapshotCache = SnapshotCache,
```

Comment:
At the cost of asking a dumb question, is it safe to shallow copy a dictionary and a list?
The low-level transaction is cloning the state it should not be able to modify the snapshot state but correct me if I'm wrong the snapshot cache and the PagerStatesAllScratchesCache  modification will reflect on to the snapshot wouldn't they?

Reply:
That is _not_ a dumb question.
And the usual answer is that it is _not_ safe.

In this case, however, the list / dictionary are readonly and are never mutated.

Reply:
Maybe use ReadOnlyDictionary<TKey, TValue> and friends to make sure this doesn't happen?

Reply:
I looked into doing that, and I worry that this may cause perf issues.
I know that `List<T>` has special treatment and is highly efficient. Turning that to `ReadOnlyCollection` may have impact that we don't like. 

I'm going to make the change anyway, because it it relatively self contained. But it is something to be aware off and we'll need to see if this pops up in the profiler output for 4.1